pkg/ebpf/perf/event.go

```diff
@@ -0,0 +1,450 @@
+	ebpfTelemetry.ReportPerfMapTelemetry(pm)
+	ebpfTelemetry.ReportPerfMapChannelLenTelemetry(pm, func() int {
+		return int(e.chLenTelemetry.Swap(0))
+	})
+	e.f = pm
+}
+
+func (e *EventHandler) perfRecordHandler(record *perf.Record, _ *manager.PerfMap, _ *manager.Manager) {
+	e.perfChan <- record
+	if e.opts.telemetryEnabled {
+		updateMaxTelemetry(e.chLenTelemetry, uint64(len(e.perfChan)))
+	}
+}
+
+func (e *EventHandler) perfLoopHandler(record *perf.Record) {
+	// record is only allowed to live for the duration of the callback. Put it back into the sync.Pool once done.
+	defer perfPool.Put(record)
+	e.handler(record.RawSample)
+}
+
+func (e *EventHandler) initRingBuffer(mgr *manager.Manager) {
+	e.ringChan = make(chan *ringbuf.Record, e.opts.channelSize)
+	e.readLoop = e.ringLoop
+
+	// remove any existing matching ring buffers from manager
+	mgr.RingBuffers = slices.DeleteFunc(mgr.RingBuffers, func(ringBuf *manager.RingBuffer) bool {
+		return ringBuf.Name == e.mapName
+	})
+	rb := &manager.RingBuffer{
+		Map: manager.Map{Name: e.mapName},
+		RingBufferOptions: manager.RingBufferOptions{
+			RecordHandler:    e.ringRecordHandler,
+			RecordGetter:     ringbufPool.Get,
+			TelemetryEnabled: e.opts.telemetryEnabled,
+		},
+	}
+	mgr.RingBuffers = append(mgr.RingBuffers, rb)
+	ebpfTelemetry.ReportRingBufferTelemetry(rb)
+	ebpfTelemetry.ReportRingBufferChannelLenTelemetry(rb, func() int {
+		return int(e.chLenTelemetry.Swap(0))
+	})
+	e.f = rb
+}
+
+func (e *EventHandler) ringLoop() {
+	for record := range e.ringChan {
+		e.ringLoopHandler(record)
+	}
+}
+
+func (e *EventHandler) ringRecordHandler(record *ringbuf.Record, _ *manager.RingBuffer, _ *manager.Manager) {
+	e.ringChan <- record
+	if e.opts.telemetryEnabled {
+		updateMaxTelemetry(e.chLenTelemetry, uint64(len(e.ringChan)))
+	}
+}
+
+func (e *EventHandler) ringLoopHandler(record *ringbuf.Record) {
+	// record is only allowed to live for the duration of the callback. Put it back into the sync.Pool once done.
+	defer ringbufPool.Put(record)
+	e.handler(record.RawSample)
+}
+
+// UpgradePerfBuffer upgrades a perf buffer to a ring buffer by creating a map spec editor
+func UpgradePerfBuffer(mgr *manager.Manager, mgrOpts *manager.Options, mapName string) {
+	if mgrOpts.MapSpecEditors == nil {
+		mgrOpts.MapSpecEditors = make(map[string]manager.MapSpecEditor)
+	}
+	specEditor := mgrOpts.MapSpecEditors[mapName]
+	specEditor.Type = ebpf.RingBuf
+	specEditor.KeySize = 0
+	specEditor.ValueSize = 0
+	specEditor.EditorFlag |= manager.EditType | manager.EditKeyValue
+	mgrOpts.MapSpecEditors[mapName] = specEditor
+
+	// remove map from perf maps because it has been upgraded
+	mgr.PerfMaps = slices.DeleteFunc(mgr.PerfMaps, func(perfMap *manager.PerfMap) bool {
+		return perfMap.Name == mapName
+	})
+}
+
+func updateMaxTelemetry(a *atomic.Uint64, val uint64) {
+	for {
+		oldVal := a.Load()
+		if val <= oldVal {
+			return
+		}
+		if a.CompareAndSwap(oldVal, val) {
+			return
+		}
+	}
+}
```

Comment:
please document the function and why you're doing it in that way

Reply:
What do you mean "that way"? This is how you would atomically update a max value

Reply:
then please document